src/services/codefixes/addMissingConstInForLoop.ts

```diff
@@ -6,18 +6,48 @@ namespace ts.codefix {
         errorCodes,
         getCodeActions: (context) => {
             const changes = textChanges.ChangeTracker.with(context, t => makeChange(t, context.sourceFile, context.span.start));
-            return [createCodeFixAction(fixId, changes, Diagnostics.Add_const_modifier_to_unresolved_variable, fixId, Diagnostics.Add_const_modifiers_to_all_unresolved_variables)];
+            if (changes) {
+                return [createCodeFixAction(fixId, changes, Diagnostics.Add_const_to_unresolved_variable, fixId, Diagnostics.Add_const_to_all_unresolved_variables)];
+            }
         },
         fixIds: [fixId],
         getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, diag) => makeChange(changes, diag.file, diag.start)),
     });
 
     function makeChange(changeTracker: textChanges.ChangeTracker, sourceFile: SourceFile, pos: number) {
-        const token = getTokenAtPosition(sourceFile, pos);
-        // fails on 'for ([x, y] of [[1,2]]) {}' when called for y
-        // since findPrecedingMatchingToken does not return the open paren here after iterating over another identifier (x)
-        const openParenToken = <Node>findPrecedingMatchingToken(token, SyntaxKind.OpenParenToken, sourceFile);
-        Debug.assert(!!openParenToken, "openParenToken must be defined");
-        changeTracker.insertNodeAt(sourceFile, openParenToken.getEnd(), createToken(SyntaxKind.ConstKeyword), { suffix: " " });
+        const forInitializer = findAncestor(getTokenAtPosition(sourceFile, pos), node =>
+            isForInOrOfStatement(node.parent) ? node.parent.initializer === node
+                : isPossiblyPartOfDestructuring(node) ? false : "quit");
+        if (!forInitializer) return;
+        if (alreadyContainsConstCodeFixForInitializer(changeTracker, forInitializer, sourceFile)) return;
+        changeTracker.insertNodeBefore(sourceFile, forInitializer, createToken(SyntaxKind.ConstKeyword));
+    }
+
+    function isPossiblyPartOfDestructuring(node: Node): boolean {
+        switch (node.kind) {
+            case SyntaxKind.Identifier:
+            case SyntaxKind.ArrayLiteralExpression:
+            case SyntaxKind.ObjectLiteralExpression:
+            case SyntaxKind.PropertyAssignment:
+            case SyntaxKind.ShorthandPropertyAssignment:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    function alreadyContainsConstCodeFixForInitializer(changeTracker: textChanges.ChangeTracker, forInitializer: Node, sourceFile: SourceFile): boolean {
+        return changeTracker.getChanges().some(change => {

```

Comment:
This will be brittle if the fix changes slightly. It would be better to keep a `NodeSet` of all the initializers and only do the change if `tryAdd(forInitializer)` succeeds.

Reply:
That sounds like a good idea. I am having trouble finding the correct scope to store this `NodeSet` in. Could you point me in the right direction, please?

Reply:
Create it at the start of `getAllCodeActions` before the `codeFixAll` loop. (Since this is irrelevant if we're only doing a single code fix.) See `disableJsDiagnostics` for an example (though that uses a set of numbers instead of a set of nodes).

Reply:
Thanks for the hint, I believe this allowed for a positive change regarding both stability and readability.